66_HLSLBxDFTests/app_resources/tests.hlsl

```diff
@@ -0,0 +1,455 @@
+#ifndef BXDFTESTS_TESTS_HLSL
+#define BXDFTESTS_TESTS_HLSL
+
+#include "nbl/builtin/hlsl/cpp_compat.hlsl"
+
+#include "nbl/builtin/hlsl/random/xoroshiro.hlsl"
+#include "nbl/builtin/hlsl/bxdf/common.hlsl"
+#include "nbl/builtin/hlsl/bxdf/reflection.hlsl"
+#include "nbl/builtin/hlsl/bxdf/transmission.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+
+using ray_dir_info_t = bxdf::ray_dir_info::SBasic<float>;
+using iso_interaction = bxdf::surface_interactions::SIsotropic<ray_dir_info_t>;
+using aniso_interaction = bxdf::surface_interactions::SAnisotropic<ray_dir_info_t>;
+using sample_t = bxdf::SLightSample<ray_dir_info_t>;
+using iso_cache = bxdf::SIsotropicMicrofacetCache<float>;
+using aniso_cache = bxdf::SAnisotropicMicrofacetCache<float>;
+using quotient_pdf_t = bxdf::quotient_and_pdf<float32_t3, float>;
+
+inline float rngFloat01(NBL_REF_ARG(nbl::hlsl::Xoroshiro64Star) rng)
+{
+    return (float)rng() / numeric_limits<uint32_t>::max;
+}
+
+inline float32_t3 rngFloat301(NBL_REF_ARG(nbl::hlsl::Xoroshiro64Star) rng)
+{
+    return float32_t3(rngFloat01(rng), rngFloat01(rng), rngFloat01(rng));
+}
+
+
+struct SBxDFTestResources
+{
+    static SBxDFTestResources create(uint32_t2 seed)
+    {
+        SBxDFTestResources retval;
+
+        retval.rng = nbl::hlsl::Xoroshiro64Star::construct(seed);
+        retval.u = float32_t3(rngFloat01(retval.rng), rngFloat01(retval.rng), 0.0);
+
+        retval.V.direction = nbl::hlsl::normalize<float32_t3>(rngFloat301(retval.rng));
+        retval.N = nbl::hlsl::normalize<float32_t3>(rngFloat301(retval.rng));
+        retval.T = nbl::hlsl::normalize<float32_t3>(rngFloat301(retval.rng));
+
+        retval.T = nbl::hlsl::normalize<float32_t3>(retval.T - nbl::hlsl::dot<float32_t3>(retval.T, retval.N) * retval.N); // gram schmidt
+        retval.B = nbl::hlsl::cross<float32_t3>(retval.N, retval.T);
+
+        retval.alpha.x = rngFloat01(retval.rng);
+        retval.alpha.y = rngFloat01(retval.rng);
+        retval.eta = 1.3;// rngFloat01(retval.rng) + 1.0;
+        retval.ior = float32_t3x2(1.02, 1.0,      // randomize at some point?
+                                1.3, 2.0,
+                                1.02, 1.0);
+        return retval;
+    }
+
+    ray_dir_info_t dV(int axis)
+    {
+        float32_t3 d = (float32_t3)0.0;
+        d[axis] += h;
+        ray_dir_info_t retval;
+        retval.direction = V.direction + d;
+        return retval;
+    }
+
+    float h = 0.001;
+
+    nbl::hlsl::Xoroshiro64Star rng;
+    ray_dir_info_t V;
+    float32_t3 N;
+    float32_t3 T;
+    float32_t3 B;
+
+    float32_t3 u;
+    float32_t2 alpha;
+    float eta;
+    float32_t3x2 ior;
+};
+
+struct STestMeta
+{
+    float32_t4 result;
+#ifndef __HLSL_VERSION
+    std::string bxdfName;
+    std::string testName;
+#endif
+};
```

Comment:
returning a `result` as a vec4 is quite a bit meh, I'd actually provide a callback (via static polymorphism) which is only called upon a failure
```cpp
struct FailureCallbackConceptExample
{
   void __call(NBL_CONST_REF_ARG(SBxDFTestResources) failedFor, NBL_CONST_REF_ARG(sample_t) failedAt);
};
```


then you can print in C++, mark booleans, append to vector, break & repeat call or whatever

Reply:
perfect C++ callback
```cpp
printf(...your error message...);
for (volatile bool repeat=true; gDebuggerAttached&&repeat; )
{
   repeat = false;
   __debugbreak();
   failedFor.test(failedAt);
}
```